CAP-035: Claimable balance clawback

sisuresh opened the pull request:
# Description

This PR implements the Claimable Balance related changes specified at https://github.com/stellar/stellar-protocol/blob/master/core/cap-0035.md.

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

jonjove commented on src/ledger/TrustLineWrapper.cpp:
```diff
@@ -377,6 +391,13 @@ TrustLineWrapper::IssuerImpl::isAuthorizedToMaintainLiabilities() const
+bool
+TrustLineWrapper::IssuerImpl::isClawbackEnabled() const
+{
+    // an issuer can't clawback from itself
```
I'm not sure about this. An issuer definitely shouldn't be able to do a normal clawback on itself (because that has no effect). But clawing back an issuer claimable balance is observable, so maybe that should work? @leighmcculloch thoughts?

leighmcculloch replied:
I agree with the comment, an issuer account can't clawback from itself. If the issuer account creates a claimable balance containing their asset, they can claw it back, the claimable balance is not the issuer account.

jonjove replied:
After talking to @sisuresh, this is trickier than I thought. Making all issuer-created claimable balances clawback enabled breaks backwards compatibility (as soon as the protocol 16 upgrade happens, this behavior will change). Making it depend on the account flag of the issuer isn't great, because the issuer might want to be able to clawback their own claimable balances without having clawback enabled for all new trust lines.

There are some ways to workaround this. Issuers could do the "add themselves as a claimant" method that we have discussed before. Alternatively, issuers could send claimable balances from a distribution account that they control which has a clawback enabled trustline.

Nothing here is really great, as far as I'm concerned. What do you think is the best approach @leighmcculloch @sisuresh?

leighmcculloch replied:
> Making all issuer-created claimable balances clawback enabled breaks backwards compatibility

Does this occur because an issuer account has no trustline from which to copy the clawback enabled state? That's interesting, I hadn't considered that.

I think we should discuss this problem on the mailing list. I cross-posted the problem here:
https://groups.google.com/g/stellar-dev/c/hPhkXhrl5-Y/m/Eg6VwKiCAQAJ

sisuresh replied:
I update the implementation so the issuer can clawback a claimable balance it creates if `AUTH_CLAWBACK_ENABLED_FLAG` is set on the issuer at the time of balance creation.

leighmcculloch replied:
I'm updating the CAP in https://github.com/stellar/stellar-protocol/pull/857.

jonjove reviewed (commented):
Overall, I think this is really good. I had one important question which is more about what the protocol _should_ do rather than what you implemented.

sisuresh commented on src/transactions/TransactionUtils.h:
```diff
@@ -163,6 +163,7 @@ bool isAuthRequired(ConstLedgerTxnEntry const& entry);
 
 bool isClawbackEnabledOnTrustline(LedgerTxnEntry const& entry);
 bool isClawbackEnabledOnAccount(ConstLedgerTxnEntry const& entry);
+bool isClawbackEnabledOnClaimableBalance(LedgerEntry const& entry);
```
I don't see `isClawbackEnabledOnClaimableBalance` used in a previous commit.

jonjove replied:
Misread something, sorry my mistake.

sisuresh commented on src/transactions/ClawbackClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,105 @@
+        return false;
+    }
+
+    switch (claimableBalance.asset.type())
```
Done

sisuresh commented on src/transactions/ClawbackClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,105 @@
+
+    switch (claimableBalance.asset.type())
+    {
+    case ASSET_TYPE_NATIVE:
```
Not sure how I didn't the duplicate check... Fixed

sisuresh commented on src/transactions/TransactionUtils.cpp:
```diff
@@ -1002,6 +1002,17 @@ getAsset(AccountID const& issuer, AssetCode const& assetCode)
+void
+setClaimableBalanceClawbackEnabled(ClaimableBalanceEntry& cb)
+{
+    if (cb.ext.v() == 0)
```
So I ended up changing the method and moving it to `CreateClaimableBalanceOpFrame`. I think it's specific enough to not require a `prepare`. This method won't be used outside of `CreateClaimableBalanceOpFrame`, so I didn't see any point in separating responsibilities. Please let me know if you disagree though.

jonjove replied:
I think this is fine, because right now this is the only way to get `cb.ext.v() == 1`. If we add another flag, then I think we'll take the `prepare` approach.

sisuresh commented on src/transactions/test/ClawbackClaimableBalanceTests.cpp:
```diff
@@ -0,0 +1,183 @@
+            gateway.clawbackClaimableBalance(balanceID);
+        }
+
+        SECTION("clawback sponsored claimable balance")
```
Added a test where the issuer is the sponsor, and another where the issuer creates a claimable balance. This also tests the new logic we agreed on where the issuer can clawback a claimable balance it creates if `AUTH_CLAWBACK_ENABLED_FLAG` is set on the issuer at the time of balance creation.

sisuresh commented on src/transactions/ClawbackClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,105 @@
+
+    auto header = ltx.loadHeader();
+    auto sourceAccount = loadSourceAccount(ltx, header);
+    removeEntryWithPossibleSponsorship(
```
That's a good point. It can definitely be improved to accommodate claimable balances. I'd say I'm at least partially to blame here since I adapted it to work for claimable balances instead of improving the interface...

jonjove replied:
Tracked in https://github.com/stellar/stellar-core/issues/2899.

sisuresh commented:
@jonjove I added fixup comments so the changes are easier to review, but forgot I rebased with upstream/master instinctively before making the changes... sorry about that.

jonjove commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -14,6 +14,20 @@
+    }
+
+    cb.ext.v(1);
+    cb.ext.v1().flags = 0;
```
Nitpick: actually could just merge these two lines (27,28).

sisuresh replied:
Done

jonjove commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -186,8 +200,13 @@ CreateClaimableBalanceOpFrame::doApply(AbstractLedgerTxn& ltx)
         }
 
+        // if the issuer is creating a claimable balance, then we need to check
+        // the account flags to detrmine if clawback should be enabled on the
```
Nitpick: typo "detrmine" -> "determine"

sisuresh replied:
Fixed

jonjove commented:
Please squash, then @MonsieurNicolas can give it a final look.

sisuresh commented:
> Please squash, then @MonsieurNicolas can give it a final look.

Squashed

MonsieurNicolas commented on src/ledger/TrustLineWrapper.cpp:
```diff
@@ -377,6 +391,13 @@ TrustLineWrapper::IssuerImpl::isAuthorizedToMaintainLiabilities() const
+TrustLineWrapper::IssuerImpl::isClawbackEnabled() const
+{
+    // an issuer can't clawback from itself
+    return false;
```
Should this be an internal error instead of `return false` ? It seems to me that it would just be a bug to reason about this flag on an issuer

sisuresh replied:
Yeah I think that's a good idea and it can be done with your suggested update to `CreateClaimableBalance`. Done.

MonsieurNicolas commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -14,6 +14,19 @@
 {
 
+void
+setClaimableBalanceClawbackEnabled(ClaimableBalanceEntry& cb)
```
this should be static/anonymous namespace as it's only used here

sisuresh replied:
Done. I added another commit that did the same for the other free functions here.

MonsieurNicolas commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -179,13 +198,19 @@ CreateClaimableBalanceOpFrame::doApply(AbstractLedgerTxn& ltx)
+        // balance
+        if (header.current().ledgerVersion >= 16 &&
+            (trustline.isClawbackEnabled() ||
+             (getSourceID() == getIssuer(asset) &&
```
This logic/comment reads backwards:
I'd expect to just read the condition as "for issuers, we use the account flag otherwise we use the trustline flag".

So something like:
```
        if (header.current().ledgerVersion >= 16)
        {
            bool enableClawback;
            if (getSourceID() == getIssuer(asset))
            {
                enableClawback = isClawbackEnabledOnAccount(sourceAccount);
            }
            else
            {
                enableClawback = trustline.isClawbackEnabled();
            }

            if (enableClawback)
            {
                setClaimableBalanceClawbackEnabled(claimableBalanceEntry);
            }
        }
```

More verbose, but easier to read/reason about and we don't need that comment anymore.

sisuresh replied:
Done.

MonsieurNicolas commented on src/ledger/test/LedgerTestUtils.cpp:
```diff
@@ -235,6 +235,11 @@ makeValid(ClaimableBalanceEntry& c)
+
+    if (c.ext.v() == 1)
+    {
+        c.ext.v1().flags &= MASK_CLAIMABLE_BALANCE_FLAGS;
```
this is not valid. `flags == 0` iif `v() == 0`. So this line should be `c.ext.v1().flags = MASK_CLAIMABLE_BALANCE_FLAGS`

sisuresh replied:
Done.

MonsieurNicolas commented on src/transactions/TransactionUtils.cpp:
```diff
@@ -1013,6 +1013,17 @@ getAsset(AccountID const& issuer, AssetCode const& assetCode)
+{
+    if (cb.ext.v() == 1)
+    {
+        return (cb.ext.v1().flags & ~MASK_CLAIMABLE_BALANCE_FLAGS) == 0;
```
in code you also assume (in `setClaimableBalanceClawbackEnabled`) that if `ext.v() == 1`, then `ext.v1().flags != 0` ; ie can't clear the flag anywhere in code. So this is really `cb.ext.v1().flags == MASK_CLAIMABLE_BALANCE_FLAGS`

sisuresh replied:
Good point. Done.

MonsieurNicolas reviewed (changes requested):
minor suggestions; looks good overall

MonsieurNicolas commented:
@sisuresh PR after those latest updates looks good to me!

jonjove reviewed (commented):
This looks good to me too. There is one minor thing to touch up, then this is ready to go.

sisuresh commented on src/transactions/ClawbackClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,86 @@
+        return false;
+    }
+
+    auto const& claimableBalance =
```
Fixed

jonjove commented:
r+ 912efa8
